[Settings] Use DataForms to render settings fields

chihsuan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Related to https://github.com/woocommerce/woocommerce/issues/52316

This PR shows how we can use DataForms to render settings fields instead of custom code. 

To use DataForm, we need to ensure each setting field has an `id` property, so I updated the setting data transformer to add a random id if it's missing.


https://github.com/user-attachments/assets/c466cbc0-0b44-47af-a3e2-a244303137b7


---

DataForm is a component that allows us to render form fields with a consistent UI. It supports a variety of field types, including text, number, checkbox, and select. Additionally, it's possible to render custom fields by providing a custom component. 

For more details, see:

- [DataForm README](https://github.com/WordPress/gutenberg/blob/227fcd87f1eccf4b853850c8f439fd95ba1e5f5f/packages/dataviews/README.md#dataform)
- [DataForm storybook](https://wordpress.github.io/gutenberg/?path=/docs/dataviews-dataform--docs&args=type:regular)

`@wordpress/dataviews@4.10.0` has a bug in package.json that causes the tests to fail. It has been fixed in https://github.com/WordPress/gutenberg/pull/67962 but not yet released so I've created a patch to fix it for now. Once it's released, we can remove the patch.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add/Enable Gutenberg version `19.8.0` or earlier.
2. Turn on the `settings` feature flag
3. Use the [Settings Tester](https://github.com/woocommerce/woocommerce-settings-migration-tester) to test Modern screens. Clone it locally and `npm install && npm run plugin-zip` / [settings-tester.zip](https://github.com/user-attachments/files/18208344/settings-tester.zip)
3. Upload and activate the Tester
4. Go to `WooCommerce > Settings` and poke around and see checkboxes, selects, and inputs rendered with values.
5. Go to `WooCommerce > Settings > Emails` and see a custom HTML table rendered.
6. Next go to Site visibility to confirm our slotFill root is available by querying the document in the console.

```js
document.getElementById('wc_settings_site_visibility_slotfill');
```

7. Go to `/wp-admin/admin.php?page=wc-settings&tab=settings-tester&section=javascript`
8. Toggle `Echoing HTML` checkbox and confirm the yellow background is shown when the checkbox is checked.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->
Use DataForms to render settings fields 

</details>

chihsuan committed:
Ensure setting item has id

chihsuan committed:
Update types

chihsuan committed:
Update components

chihsuan committed:
Fix test

chihsuan committed:
Update styles

chihsuan committed:
Use @wordpress/dataviews ^4.10.0

chihsuan committed:
Fix types

chihsuan committed:
Update pnpm-lock.yaml

chihsuan committed:
Fix tests

Commit:
Add changefile(s) from automation for the following project(s): @woocommerce/settings-editor, woocommerce

chihsuan committed:
Update package.json and pnpm-lock.yaml to include patch for @wordpress/dataviews@4.10.0 and add new patch file

chihsuan committed:
Update custom view id

chihsuan committed:
Rename file and add docs

chihsuan committed:
Update transformer

psealock commented on packages/js/settings-editor/src/utils/dataforms-transformers.tsx:
```diff
@@ -0,0 +1,227 @@
+						value,
+					} )
+				),
+				Edit: 'select',
```
Maybe its a problem upstream but I'm seeing this error:
```
react-dom.js?ver=18:73 Warning: The `value` prop supplied to <select> must be a scalar value if `multiple` is false.
```

psealock replied:
Also an upstream problem, but this select does not apply the id to the element, instead using something like `inspector-select-control-1`. We could use `SelectControl` in a similar way that you've used `InputControl` in this PR. 

It will be important for input elements to have the `id` applied because the legacy forms follow this pattern. Some extensions assume this is the case for things like form manipulation or even creating their own submit button.

Alternatively we could make this contribution upstream.

chihsuan replied:
Good catch! I'll add a SelectControl.

chihsuan replied:
Added in 7bd2444ef4968e6db4458ac587d0acfbb3df6f1f.

psealock commented on packages/js/settings-editor/src/utils/dataforms-transformers.tsx:
```diff
@@ -0,0 +1,227 @@
+	fields: settings
+		.map( transformToFormField )
+		.filter( ( field ) => field !== false ),
+} );
```
We'll need to add some js unit tests for these functions 😄

psealock reviewed (commented):
This is really awesome @chihsuan !

This is working well, I just noted issues with ids and select.

>DataForm is a component that allows us to render form fields with a consistent UI. It supports a variety of field types, including text, number, checkbox, and selects

The consistency in UI is great, but given that we will be passing in `InputControl`, possibly `SelectControl` and a few others do you think the added complexity is worthwhile? One thought I have is this depends on what changes we can get merged upstream. For instance, if can get `InputControl` rendered by dataviews instead of what it does now, this will support several more text types (email, password, etc) out of the box and we can remove the `getInputEdit` you've added here. This example seems like a no-brainer, but I was curious if you have any further thoughts on this?

chihsuan committed:
Update pnpm-lock.yaml

kalessil commented on package.json:
```diff
@@ -77,7 +77,8 @@
 		"patchedDependencies": {
-			"@wordpress/edit-site@5.15.0": "bin/patches/@wordpress__edit-site@5.15.0.patch"
+			"@wordpress/edit-site@5.15.0": "bin/patches/@wordpress__edit-site@5.15.0.patch",
+			"@wordpress/dataviews@4.10.0": "patches/@wordpress__dataviews@4.10.0.patch"
```
It looks like the new patch location needs to be updated. We keep patches under `bin/patches/` directory.

chihsuan replied:
Ah, that's right. Thanks for catching this! I just noticed that `@wordpress/dataviews@4.11.1` has been released, which includes the fix for the issue so I'll update the dependency and remove the patch.

chihsuan replied:
Updated in fc4e1498112a92a76e51d3faf65e301ed2f011de.

kalessil replied:
Super, thanks!

chihsuan committed:
Use @wordpress/dataviews 4.11.1 and remove the patch

chihsuan commented:
> The consistency in UI is great, but given that we will be passing in InputControl, possibly SelectControl and a few others do you think the added complexity is worthwhile?
>This example seems like a no-brainer, but I was curious if you have any further thoughts on this?


@psealock Yeah, as discussed in the meetup, I think we could contribute upstream so that we don't need to implement our custom InputControl, SelectControl, etc (though we're pausing on contributing upstream now 😅 ).

chihsuan committed:
Add SelectEdit component and integrate into dataforms transformers

chihsuan committed:
Delete components

chihsuan committed:
Add tests

chihsuan committed:
Remove settings item

chihsuan committed:
Rename

chihsuan committed:
Update pnpm-lock.yaml

chihsuan committed:
Create useSettingsForm hook

chihsuan committed:
Remove unsupported types

chihsuan committed:
Import dataviews style

chihsuan commented:
@psealock  I've addressed the issues you mentioned. I also added a `useSettingsForm` hook to make it easier to understand. Could you please review it again? 🙏  Thanks!

kalessil reviewed (approved):
LGTM (I reviewed only patch/deps bump part)!

louwie17 commented:
Just a drive by comment here :)
Nice work @chihsuan, and happy to see you got this also working with DataForm's. I see in some previous comments you mentioned around making changes upstream. Given the pause, it may be good in the mean time to surface some of your feedback either by creating an issue in Gutenberg, sharing some feedback in [this Slack channel](https://wordpress.slack.com/archives/C082S8Q9PL3), or on P2 ( internally ). 

Btw I saw the big switch statement, we currently have something similar within the DataForm package: https://github.com/WordPress/gutenberg/blob/0ceac8f5e8b74cbeb7ef36c44009c67ec5fabbf7/packages/dataviews/src/field-types/index.tsx#L15. The plan is to elaborate on this, for example your `getInputEdit` solution may be a great addition to this. We also want to create some functions around this to be able to register field types, so you don't have to explicitly these definitions through the `fields` list.

Anyhow glad to see you make it work and the more use cases the better :)

psealock reviewed (approved):
This is looking great, code looks good and tests well 🚢

moon0326 reviewed (approved):
LGTM and tested well 👍 🚀

psealock commented:
One thing I just noticed is when a text field has no `desc` then the label is just the id. For `text` types we should also respect `placeholder` values and figure out what to do with `desc_tip`. We could address this in a follow up or just create an issue for the backlog.

<img width="753" alt="image" src="https://github.com/user-attachments/assets/c3e82fc3-2be3-49ec-a810-c4c4fada9e03" />

chihsuan commented:
> One thing I just noticed is when a text field has no `desc` then the label is just the id. For `text` types we should also respect `placeholder` values and figure out what to do with `desc_tip`. We could address this in a follow up or just create an issue for the backlog.

Thanks, I'll create an issue for this.
